File: leetcode/leetcode_cpp/longest-palindromic-substring.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include "../utils.h"

using namespace std;

/**
https://leetcode.com/problems/longest-palindromic-substring/

// solution1. expand from center
time: o(n^2) : iterate(n) expand(n)
space: o(1)

babad
*         b
 *        a
  *       aba
  l
  r
 l r
   *.     a
    *.    d
    
babbad
  *       abba
  lr
 l  r

// solution2. manacher's al

*/

class Solution {
public:
    string longestPalindrome_1(string& s) {
        if (s.size() < 2) return s;
        int maxLength = 1;
        int start = 0;
        for (int i=0; i<s.size(); ++i) {
            int l = i;
            while (i < s.size()-1 && s[i+1] == s[l]) i++;

            int r = i;
            while (l >= 0 && r<s.size() && s[l] == s[r]) {
                int length = r - l + 1;
                if (maxLength < length) {
                    start = l;
                    maxLength = length;
                }
                l--;
                r++;
            }
            if (r == s.size() - 1) break;
        }
        return s.substr(start, maxLength);
    }
};

int main()
{
    assert(Solution().longestPalindrome(
        "babad"
    ) == ("bab"));
    assert(Solution().longestPalindrome(
        "cbbd"
    ) == ("bb"));
    return 0;
}

File: leetcode/leetcode_py/longest-palindromic-substring.py
'''
https://leetcode.com/problems/longest-palindromic-substring/

Solution1. Iterate and expand from cetner
time: o(n^2)
space: o(1)

Solution2. DP (https://leetcode.com/problems/longest-palindromic-substring/solutions/147548/direct-c-dp/?envType=study-plan-v2&envId=top-interview-150)
time: o(n^2)
space: o(n^2)
'''
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        if n < 2: return s
        maxRange = 1
        start = 0
        for i in range(n):
            l = i
            while i < n-1 and s[l] == s[i+1]: i += 1
            r = i
            while l >= 0 and r < n and s[l] == s[r]:
                length = r - l + 1
                if maxRange < length:
                    print(l, r, length)
                    maxRange = length
                    start = l
                l -= 1
                r += 1
            if r == n - 1: break
        return s[start:start+maxRange]
                
            

        

            
        